docs/en/setup/configuration/config.md
# configuration

## Logging

Configure system logs in SkyWalking Predictor.

| Name           | Default                                               | Environment Key | Description               |
|----------------|-------------------------------------------------------|-----------------|---------------------------|
| logging.level  | INFO                                                  | LOGGING_LEVEL   | Minimum log output level. |
| logging.format | :%(asctime)s - %(name)s - %(levelname)s - %(message)s | LOGGING_FORMAT  | Log output format.        |

## server

Configure the external services provided by SkyWalking Predictor.

| Name                   | Default | Environment Key | Description                                              |
|------------------------|---------|-----------------|----------------------------------------------------------|
| server.grpc.port       | 18080   | GRPC_PORT       | Port for providing external gRPC services.               |
| server.monitor.enabled | true    | MONITOR_ENABLED | Whether to enable Prometheus metrics monitoring service. |
| server.monitor.port    | 8000    | MONITOR_PORT    | Port for providing external monitoring services.         |

## baseline

Configure how to fetch data, and prediction metrics for dynamic baseline.

Please make these module in OAP have been activated:
1. **status-query**: Query `/status/config/ttl` for getting TTL of days for fetch all metrics data.
2. **graph** in **query**: Query service, metrics from GraphQL.

| Name                                | Default                 | Environment Key                     | Description                                                                                                                                             |
|-------------------------------------|-------------------------|-------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------|
| baseline.cron                       | */8 * * * *             | BASELINE_FETCH_CRON                 | Configure the execution timing of data retrieval and prediction for the baseline by a cron expression.                                                  |
| baseline.fetch.server.address       | http://localhost:12800/ | BASELINE_FETCH_SERVER_ENDPOINT      | Address of OAP Restful server.                                                                                                                          |
| baseline.fetch.server.username      |                         | BASELINE_FETCH_SERVER_USERNAME      | If OAP access requires authentication, the username must be provided.                                                                                   |
| baseline.fetch.server.password      |                         | BASELINE_FETCH_SERVER_USERNAME      | If OAP access requires authentication, the password must be provided.                                                                                   |
| baseline.fetch.server.down_sampling | HOUR                    | BASELINE_FETCH_SERVER_DOWN_SAMPLING | Specify the type of downsampling data to download from OAP, supporting `HOUR` and `MINUTE`. Note that retrieving minute-level data takes a longer time. |
| baseline.fetch.server.layers        | GENERAL                 | BASELINE_FETCH_SERVER_LAYERS        | Specify which layer service data needs to be fetch. Use a comma(`,`) to separate multiple layers.                                                       |
| baseline.fetch.metrics              |                         |                                     | List of metrics to be monitored.                                                                                                                        |
| baseline.fetch.metrics.name         |                         |                                     | Metric name.                                                                                                                                            |
| baseline.fetch.metrics.enabled      |                         |                                     | Is active the Metric or not.                                                                                                                            |
| baseline.fetch.predict.directory    | ./out_predict           | BASELINE_PREDICT_DIRECTORY          | The directory for save prediction results for query purposes.                                                                                           |
| baseline.fetch.predict.min_days     | 2                       | BASELINE_PREDICT_MIN_DAYS           | The minimum number of days of data required for metric prediction, preventing inaccuracies due to insufficient data.                                    |
| baseline.fetch.predict.frequency    | h                       | BASELINE_PREDICT_FREQUENCY          | Specify the frequency of the predicted data. Currently, only hourly (`h`) is supported.                                                                 |
| baseline.fetch.predict.period       | 24                      | BASELINE_PREDICT_PERIOD             | Specify the number of future data points to predict.                                                                                                    |



docs/en/setup/configuration/override-settings.md
# Setting Override
SkyWalking Predictor supports setting overrides by system environment variables. 
You could override the settings in `config.yaml`

## System environment variables
- Example

  Override `server.grpc.port` in this setting segment through environment variables
  
```yaml
server:
  grpc:
    port: "${GRPC_PORT:18080}"
```

If the `GRPC_PORT ` environment variable exists in your operating system and its value is `9999`, 
then the value of `server.grpc.port` here will be overwritten to `9999`, otherwise, it will be set to `18080`.

baseline/config/config.py
#  Copyright 2025 SkyAPM org
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#  http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import os
import re
from typing import List, Optional

import yaml
from pydantic import BaseModel, model_validator

current_dir = os.path.dirname(os.path.realpath(__file__))

env_regular_regex = re.compile(r'\${(?P<ENV>[_A-Z0-9]+):(?P<DEF>.*)}')


class LogConfig(BaseModel):
    level: str = 'INFO'
    format: str = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'


class MonitorConfig(BaseModel):
    enabled: bool = False
    port: int = 8000


class GRPCConfig(BaseModel):
    port: int = 18080


class ServerConfig(BaseModel):
    grpc: GRPCConfig
    monitor: MonitorConfig


class BaselineFetchMetricsTagMappingConfig(BaseModel):
    key: str
    value: str


class BaselineFetchValuePreProcessConfig(BaseModel):
    hour_to_minute: bool = False


class BaselineFetchMetricsConfig(BaseModel):
    name: str
    enabled: bool
    pre_process: Optional[BaselineFetchValuePreProcessConfig] = None


class BaselineFetchGraphqlServerConfig(BaseModel):
    address: str
    username: Optional[str]
    password: Optional[str]
    down_sampling: str = "HOUR"
    layers: List[str]

    @model_validator(mode="before")
    @classmethod
    def convert_layers(cls, values):
        if isinstance(values.get("layers"), str):
            values["layers"] = [layer.strip() for layer in values["layers"].split(",")]
        return values


class BaselineFetchConfig(BaseModel):
    server: BaselineFetchGraphqlServerConfig
    metrics: List[str]

    @model_validator(mode="before")
    @classmethod
    def convert_layers(cls, values):
        if isinstance(values.get("metrics"), str):
            values["metrics"] = [metrics.strip() for metrics in values["metrics"].split(",")]
        return values


class BaselinePredictConfig(BaseModel):
    directory: str = "/tmp"
    min_days: int = 3
    frequency: str = 'h'
    period: int = 24


class BaselineConfig(BaseModel):
    cron: str = "0 0 * * *"
    fetch: BaselineFetchConfig
    predict: BaselinePredictConfig


def parse_env_variables(value):
    if isinstance(value, str):
        match = env_regular_regex.match(value)
        if match:
            env_var, default_value = match.groups()
            return os.getenv(env_var, default_value)
    return value


def parse_nested_values(values):
    if isinstance(values, dict):
        for key, val in values.items():
            if isinstance(val, str):
                values[key] = parse_env_variables(val)
            elif isinstance(val, dict):
                values[key] = parse_nested_values(val)
            elif isinstance(val, list):
                values[key] = [parse_nested_values(item) if isinstance(item, dict) else parse_env_variables(item) for
                               item in val]
    return values


class Config(BaseModel):
    logging: LogConfig
    server: ServerConfig
    baseline: BaselineConfig

    @model_validator(mode="before")
    def replace_env_variables(cls, values):
        return parse_nested_values(values)


with open(os.path.join(current_dir, "config.yaml"), "r") as f:
    yamlLoadedConfig = yaml.safe_load(f)

current_config = Config(**yamlLoadedConfig)
